File: 프로그래머스/3/67259. ［카카오 인턴］ 경주로 건설/［카카오 인턴］ 경주로 건설.py
from collections import deque

from collections import deque

def solution(board):
    n = len(board)
    # 방향별로 비용을 따로 저장
    # costs[x][y][direction] 형태로 저장
    costs = [[[float('inf')] * 4 for _ in range(n)] for _ in range(n)]
    for i in range(4):
        costs[0][0][i] = 0
    
    directions = [(0, 1), (1, 0), (0, -1), (-1, 0)]  # 동남서북
    q = deque([(0, 0, 0, 0), (0, 0, 1, 0)])  # (x, y, direction, cost)
    
    while q:
        x, y, direction, cost = q.popleft()
        
        if cost > costs[x][y][direction]:
            continue
            
        for new_dir, (dx, dy) in enumerate(directions):
            nx, ny = x + dx, y + dy
            
            if 0 <= nx < n and 0 <= ny < n and board[nx][ny] == 0:
                new_cost = cost + 100
                if direction != new_dir:
                    new_cost += 500
                

                if new_cost < costs[nx][ny][new_dir]:
                    costs[nx][ny][new_dir] = new_cost
                    q.append((nx, ny, new_dir, new_cost))
    

    return min(costs[n-1][n-1])

File: 프로그래머스/3/67259. ［카카오 인턴］ 경주로 건설/［카카오 인턴］ 경주로 건설.java

import java.util.*;

class Solution {
    static int[] dx = {-1, 1, 0, 0}; // 상 하 좌 우
    static int[] dy = {0, 0, -1, 1};
    public int solution(int[][] board) {
        int answer = Integer.MAX_VALUE;
        int n = board.length;
        
        int[][][] cost = new int[n][n][4];
        
        for (int[][] c : cost)
            for (int[] r : c)
                Arrays.fill(r, Integer.MAX_VALUE);
        
        Deque<int[]> q = new ArrayDeque<>();
        
        q.add(new int[]{0, 0, 1, 0});
        q.add(new int[]{0, 0, 3, 0});
        cost[0][0][1] = 0;
        cost[0][0][3] = 0;
        
        while(!q.isEmpty()) {
            int[] now = q.poll();
            
            int cx = now[0], cy = now[1], dir = now[2], c = now[3];
            
            if (cx == n-1 && cy == n-1) {
                answer = Math.min(c, answer);
                continue;
            }
            
            for (int d = 0; d < 4; d++) {
                int nx = dx[d] + cx;
                int ny = dy[d] + cy;
                
                if (nx >= n || ny >= n || nx < 0 || ny < 0) continue;
                if (board[nx][ny] == 1) continue; 
                
                int nextCost = c + 100;
                if (dir != d) nextCost += 500;
                
                if (cost[nx][ny][d] > nextCost) {
                    cost[nx][ny][d] = nextCost;
                    q.add(new int[]{nx, ny, d, nextCost});
                }
            
                
            }
        }
        return answer;
    }
}